HW2/P2/parallel_vector.pyx

```diff
@@ -51,65 +51,99 @@ cdef void free_N_locks(int N, omp_lock_t *locks) nogil:
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+							np.int32_t[:] src,
+							np.int32_t[:] dest,
+							int repeat):
+	cdef:
+		int idx, r, a, b
+		omp_lock_t *locks = get_N_locks(counts.shape[0])
+
+	##########
+	# Your code here
+	# Use parallel.prange() and a lock for each element of counts to parallelize
+	# data movement.  Be sure to avoid deadlock, and double-locking.
+	##########
+	with nogil:
+		for r in range(repeat):
+
+			for idx in prange(src.shape[0], num_threads=4, schedule=static):
+
+				# to prevent deadlock, sort indices and always acquire index with smallest ID first
+				a = src[idx]  if src[idx] < dest[idx] else dest[idx]
+				b = dest[idx] if src[idx] < dest[idx] else src[idx]
+
+				# to prevent double-locking, only lock a and b if a != b, otherwise only lock a
+				if (counts[src[idx]] > 0): 
+					
+					acquire(&( locks[ a ] ))
+					if a != b:
+						acquire(&( locks[ b ] ))
+
+					counts[dest[idx]] += 1
+					counts[src[idx]]  -= 1
+
+					release(&( locks[ a ] ))
+					if a != b:
+						release(&( locks[ b ] ))
+
+	free_N_locks(counts.shape[0], locks)
 
 
 cpdef move_data_medium_grained(np.int32_t[:] counts,
-                               np.int32_t[:] src,
-                               np.int32_t[:] dest,
-                               int repeat,
-                               int N):
-   cdef:
-       int idx, r
-       int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
-       omp_lock_t *locks = get_N_locks(num_locks)
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+							   np.int32_t[:] src,
+							   np.int32_t[:] dest,
+							   int repeat,
+							   int N):
+	cdef:
+		int idx, r, a, b, a_bin, b_bin
+		int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
+		omp_lock_t *locks = get_N_locks(num_locks)
+
+	##########
+	# Your code here
+	# Use parallel.prange() and a lock for every N adjacent elements of counts
+	# to parallelize data movement.  Be sure to avoid deadlock, as well as
+	# double-locking.
+	##########
+	with nogil:
+		for r in range(repeat):
+
+			for idx in prange(src.shape[0], num_threads=4):
+
+				# to prevent deadlock, sort indices and always acquire index with smallest ID first
+				a = src[idx]  if src[idx] < dest[idx] else dest[idx]
```

Comment:
what if src[idx] or dest[idx] changes between now and when you acquire the lock?